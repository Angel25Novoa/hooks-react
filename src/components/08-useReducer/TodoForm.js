import React from 'react'
import useForm from '../../hooks/useForm'

const TodoForm = ({handleAddTodo}) => {

  const [{ description }, handleInputChange, reset] = useForm({
    description: ''
  })

  const handleSubmit = (e) => {
    e.preventDefault()

    if(description.trim().length <= 1) {
      return
    }
    
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    }

    handleAddTodo(newTodo)
    reset()
  }

  return (
    <>
      <h4>Add TO-DO</h4>
      <hr />
      <form onSubmit={ handleSubmit }>
        <input 
          className='form-control'
          type="text"
          name='description'
          placeholder='DescripciÃ³n'
          autoComplete='off'
          value={ description }
          onChange={ handleInputChange }
        />
        <button 
          type='submit'
          className='btn btn-info mt-1 btn-block'
        >
          Add
        </button>
      </form>
    </>
  )
}

export default TodoForm