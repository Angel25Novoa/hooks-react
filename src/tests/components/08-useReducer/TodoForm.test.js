import React from 'react'
import {shallow} from 'enzyme'
import TodoForm from '../../../components/08-useReducer/TodoForm'

describe((`TodoForm component tests`), () => {
  const handleAddTodo = jest.fn()
  const wrapper = shallow(<TodoForm handleAddTodo={handleAddTodo}/>)
  test((`TodoForm component should render without errors`), () => {
    expect(wrapper).toMatchSnapshot()
  })
  test((`TodoForm component should NOT call handleAddTodo on submit`), () => {
    const formSubmit = wrapper.find(`form`).prop(`onSubmit`)
    formSubmit({preventDefault: () => {}})
    wrapper.find(`form`).simulate(`submit`, {preventDefault: () => {}})
    expect(handleAddTodo).toHaveBeenCalledTimes(0)
  })
  test((`TodoForm component should call handleAddTodo on submit`), () => {
    const value = `A new todo`
    wrapper.find(`input`).prop(`onChange`)({target: {value, name: `description`}})
    const formSubmit = wrapper.find(`form`).prop(`onSubmit`)
    formSubmit({preventDefault: () => {}})
    expect(handleAddTodo).toHaveBeenCalledTimes(1)
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object))
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      desc: value,
      done: false
    })
  })
  test((`TodoForm component should reset the input value on submit`), () => {
    const value = `A new todo`
    wrapper.find(`input`).prop(`onChange`)({target: {value, name: `description`}})
    const formSubmit = wrapper.find(`form`).prop(`onSubmit`)
    formSubmit({preventDefault: () => {}})
    expect(wrapper.find(`input`).prop(`value`)).toBe(``)
  })
})