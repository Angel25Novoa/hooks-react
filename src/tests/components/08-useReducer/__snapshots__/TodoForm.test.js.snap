// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm component tests TodoForm component should render without errors 1`] = `
<Fragment>
  <h4>
    Add TO-DO
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="DescripciÃ³n"
      type="text"
      value=""
    />
    <button
      className="btn btn-info mt-1 btn-block"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
